matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/session/identity/DefaultIdentityService.kt

```diff
@@ -0,0 +1,345 @@
+                }
+
+                // Try to get a token
+                val token = getNewIdentityServerToken(urlCandidate)
+
+                identityStore.setUrl(urlCandidate)
+                identityStore.setToken(token)
+                updateIdentityAPI(urlCandidate)
+
+                updateAccountData(urlCandidate)
+            }
+            urlCandidate
+        }
+    }
+
+    private suspend fun updateAccountData(url: String?) {
+        // Also notify the listener
+        withContext(coroutineDispatchers.main) {
+            listeners.toList().forEach { tryThis { it.onIdentityServerChange() } }
+        }
+
+        updateUserAccountDataTask.execute(UpdateUserAccountDataTask.IdentityParams(
+                identityContent = IdentityServerContent(baseUrl = url)
+        ))
+    }
+
+    override fun lookUp(threePids: List<ThreePid>, callback: MatrixCallback<List<FoundThreePid>>): Cancelable {
+        if (threePids.isEmpty()) {
+            callback.onSuccess(emptyList())
+            return NoOpCancellable
+        }
+
+        return GlobalScope.launchToCallback(coroutineDispatchers.main, callback) {
+            lookUpInternal(true, threePids)
+        }
+    }
+
+    override fun getShareStatus(threePids: List<ThreePid>, callback: MatrixCallback<Map<ThreePid, SharedState>>): Cancelable {
+        if (threePids.isEmpty()) {
+            callback.onSuccess(emptyMap())
+            return NoOpCancellable
+        }
+
+        return GlobalScope.launchToCallback(coroutineDispatchers.main, callback) {
+            val lookupResult = lookUpInternal(true, threePids)
+
+            threePids.associateWith { threePid ->
+                // If not in lookup result, check if there is a pending binding
+                if (lookupResult.firstOrNull { it.threePid == threePid } == null) {
+                    if (identityStore.getPendingBinding(threePid) == null) {
+                        SharedState.NOT_SHARED
+                    } else {
+                        SharedState.BINDING_IN_PROGRESS
+                    }
+                } else {
+                    SharedState.SHARED
+                }
+            }
+        }
+    }
+
+    private suspend fun lookUpInternal(canRetry: Boolean, threePids: List<ThreePid>): List<FoundThreePid> {
+        ensureToken()
+
+        return try {
+            identityBulkLookupTask.execute(IdentityBulkLookupTask.Params(threePids))
+        } catch (throwable: Throwable) {
+            // Refresh token?
+            when {
+                throwable.isInvalidToken() && canRetry -> {
+                    identityStore.setToken(null)
+                    lookUpInternal(false, threePids)
+                }
+                throwable.isTermsNotSigned()           -> throw IdentityServiceError.TermsNotSignedException
+                else                                   -> throw throwable
+            }
+        }
+    }
+
+    private suspend fun ensureToken() {
```

Comment:
@Throws

Reply:
Not sure it is relevant. Or we should add this annotation to pretty every methods in the SDK